onnxruntime/core/providers/nuphar/scripts/model_quantizer.py

```diff
@@ -0,0 +1,317 @@
+    preset_qcfgs = {'asymm8_param0_input1' : {'W' : dict(QuantizeConfig(signed=1, reserved_bits=0, type_bits=8)),
+                                              'X' : dict(QuantizeConfig(signed=0, reserved_bits=1, type_bits=8)),
+                                              'Symmetric' : 0},
+                    'symm16_param3_input3' : {'W' : dict(QuantizeConfig(signed=1, reserved_bits=3, type_bits=16)),
+                                              'X' : dict(QuantizeConfig(signed=1, reserved_bits=3, type_bits=16)),
+                                              'Symmetric' : 1}}
+    default_qcfg = preset_qcfgs[preset_str]
+    in_mp = onnx.load(input_model)
+
+    qcfg_dict = {}
+    if qcfg_json and not export_qcfg_json:
+        with open(qcfg_json, 'r') as f:
+            qcfg_dict = json.load(f)
+
+    out_mp = onnx.ModelProto()
+    out_mp.CopyFrom(in_mp)
+    out_mp.ir_version = 5 # update ir version to avoid requirement of initializer in graph input
+    for opset in out_mp.opset_import:
+        if opset.domain == '' or opset.domain == 'onnx':
+            opset.version = 10 # bump up to opset 10, which is required for MatMulInteger
+    ms_opset = out_mp.opset_import.add()
+    ms_opset.domain = 'com.microsoft'
+    ms_opset.version = 1
+    out_mp.graph.ClearField('node')
+    remove_initializers_from_inputs(out_mp)
+    nf = NodeFactory(out_mp.graph)
+    converted_weights = {} # remember MatMul weights that have been converted, in case of sharing
+    for in_n in in_mp.graph.node:
+        if in_n.op_type == 'Slice' and len(in_n.input) == 1:
+            upgrade_slice_op(nf, in_n)
+            continue
+
+        if in_n.op_type == 'MatMul' and not only_for_scan:
+            if quantize_matmul_2d_with_weight(in_n, in_mp.graph, nf, converted_weights, qcfg_dict, export_qcfg_json, default_qcfg):
+                continue
+
+        out_n = out_mp.graph.node.add()
+        out_n.CopyFrom(in_n)
+        if in_n.op_type == 'Scan':
+            in_body = [attr for attr in in_n.attribute if attr.name == 'body'][0]
+            out_body = [attr for attr in out_n.attribute if attr.name == 'body'][0]
+            out_body.g.ClearField('node')
+            scan_nf = NodeFactory(out_mp.graph, out_body.g)
+
+            for in_sn in in_body.g.node:
+                if in_sn.op_type == 'MatMul':
+                    if quantize_matmul_2d_with_weight(in_sn, in_body.g, scan_nf, converted_weights, qcfg_dict, export_qcfg_json, default_qcfg):
+                        continue
+
+                if in_sn.op_type == 'Slice' and len(in_sn.input) == 1:
+                    upgrade_slice_op(scan_nf, in_sn)
+                    continue
+
+                out_sn = out_body.g.node.add()
+                out_sn.CopyFrom(in_sn)
+
+    onnx.save(out_mp, output_model)
+    if export_qcfg_json:
+        with open(qcfg_json, 'w') as f:
+            f.write(json.dumps(qcfg_dict, indent=2))
+
+def parse_arguments():
+  parser = argparse.ArgumentParser()
```

Comment:
Same question here. Is there something somewhere that describes what this script does, why it is needed, and when it should be used?